### PR TITLE
Vit models

### PR DESCRIPTION
**Pull request by alzaia**
This PR aims to add the following:

Main contributions:

- [X] Adds notebook for finetuning pretrained HuggingFace vision transformer (ViT) on the Office31 dataset.
- [x] Adds ViT baseline implementation.
- [X] Adds MNIST dataloader for quick tests and sanity checks of baselines.
- [X] Updates Office31 dataloader and related files
- [X] Adds notebook for training a simple CNN architecture from scratch (end-to-end) on Office31 dataset.
- [x] Adds the simple CNN model implementation.

Extra:

- [X] Some testing functions for dataloaders
- [X] Some cleaning up of files & refactoring

### CODE REVIEW

File: image_classification_simulation/data/office31_loader.py [hunk 1]
```diff
@@ -45,17 +48,75 @@ def __init__(
...
-        )
+        if "train_test_split" in hyper_params:
+            self.train_test_split = hyper_params["train_test_split"]
+        else:
+            self.train_test_split = 0.15
+
+        if hasattr(self, "model_type"):
+            if self.model_type == "vit":
+
+                feature_extractor = ViTFeatureExtractor.from_pretrained(
```
Comment (SinaSarparast) [inline image_classification_simulation/data/office31_loader.py:62]: I think we should not load the model here. It will be better to get these attributes elsewhere and pass them to the dataloader.

File: examples/classic_cnn/config.yaml [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+optimizer: adamW
+loss: CrossEntropyLoss
+patience: 5
+lr: 0.0003
+architecture: classic-cnn
+pretrained: False
+max_epoch: 3
+exp_name: cnn_test_exp
+# dataset
+data: MNIST
```
Comment (SinaSarparast) [inline examples/classic_cnn/config.yaml:12]: Can we train this CNN baseline on Office31 dataset.

File: image_classification_simulation/data/mnist_loader.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+            print("Number of workers set to:", self.num_workers)
+
+        if "image_size" in hyper_params:
+            self.image_size = hyper_params["image_size"]
+        else:
+            self.image_size = 28
+            print("image size set to:", self.image_size)
+
+        self.train_set_transformation = transforms.Compose(
+            [transforms.ToTensor(), transforms.Normalize((0.1307,), (0.3081,))]
```
[the remark this answers is not in the document]
Comment (alzaia) [reply image_classification_simulation/data/mnist_loader.py:51]: Its the standard mnist preprocessing values that are used for the mnist dataloader (probably using the mean and std of the dataset).

File: image_classification_simulation/data/office31_loader.py [hunk 4]
```diff
@@ -62,16 +63,25 @@ def __init__(
         else:
             self.image_size = 300
 
+        if "model" in hyper_params:
+            self.model_type = hyper_params["model"]
```
[the remark this answers is not in the document]
Comment (alzaia) [reply image_classification_simulation/data/office31_loader.py:67]: Good point, not anymore because I removed references to models in the dataloader. Going to remove it now

File: image_classification_simulation/models/classic_cnn_baseline.py [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+        Returns
+        -------
+        int
+            The accuracy score.
+        """
+        probs = nn.functional.softmax(logits, 0)
+        preds = torch.argmax(probs, 1)
+        return (preds == targets).sum().item() / len(targets)
+
+    def training_step(
```
Comment (SinaSarparast) [inline image_classification_simulation/models/classic_cnn_baseline.py:154]: Why the training_step is overwritten? Has there been any changes compared to the one in the BaseModel?

File: image_classification_simulation/models/vit_baseline.py [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+        -------
+        typing.Any
+            returns loss and logit scores.
+        """
+        input_data, targets = batch
+        logits = self(input_data)  # calls the forward pass of the model
+        loss = self.loss_fn(logits, targets)
+        return loss, logits
+
+    def compute_accuracy(
```
Comment (SinaSarparast) [inline image_classification_simulation/models/vit_baseline.py:60]: We use the accuracy method often. Perhaps we should add it to the BaseModel instead of keeping versions of it in different places.

Comment (alzaia) [reply image_classification_simulation/models/vit_baseline.py:60]: Yes, we could add this as an issue for next PR?

File: examples/classic_cnn/config.yaml (hunk 2 above)
Comment (alzaia) [reply examples/classic_cnn/config.yaml:12]: I did play with it on the notebook (where it gave me a 40% accuracy after 5 epochs)

File: image_classification_simulation/models/classic_cnn_baseline.py (hunk 5 above)
Comment (alzaia) [reply image_classification_simulation/models/classic_cnn_baseline.py:154]: Probably won't change but I thought we might add specific stuff in the future so easier to already have it there too. Should I remove it?
